### ISSUE TITLE
Can't fetch public key of another user

### ISSUE THREAD

**Issue by sosnovsky** [FlowCrypt/flowcrypt-ios#968]
I have test user with configured FlowCrypt app (for example `user1@test.com`). 
On Settings -> Keys screen I can see Public Key, and user can send encrypted mails.

But when I try to reply to mail sent by this user I get `One or more of your recipients are missing a public key` error, because `https://flowcrypt.com/attester/pub/user1@test.com` responds with 404 error.

@tomholub do you know why it happens? Maybe I'm doing something wrong?

**Comment by sosnovsky** [comment]
Or `https://flowcrypt.com/attester/pub/..` endpoint works only for `...@flowcrypt.com` emails, and I should manually add pub keys for other users?

**Comment by tomholub** [comment]
Could you share the actual user that you are testing this with? (over email if you don't want to share here)

**Comment by sosnovsky** [comment]
> Could you share the actual recipient user that you are testing this with? (over email if you don't want to share here)

Sent over email

**Comment by tomholub** [comment]
`https://flowcrypt.com/attester/pub/..` generally contains public keys of all FlowCrypt users except enterprise users who use WKD. Additionally, some enterprise users have attester disabled using https://flowcrypt.com/docs/business/org-rules.html `disallow_attester_search_for_domains` where they either enter domains to disable, or possibly `*` to disable Attester search altogether.

For example, here is our own org-rules for users on flowcrypt.com domain (which is set up similar to some of our enterprise customers) https://fes.flowcrypt.com/api/v1/client-configuration?domain=flowcrypt.com (we don't happen to use the `disallow_attester_search_for_domains` at the moment).

But none of that explains the issue you are having.

**Comment by tomholub** [comment]
Re-reading your original issue, I understood now that the situation is:
1) set up an account A
2) send email to someone
3) someone tries to reply, and cannot find your public key, because it's not on Attester

Here it is clear - there was a (possible) issue in step 1. That user's public key is supposed to be sent to Attester, but it wasn't. So it's not there.

Did you use the iOS app to set that user app?

**Comment by sosnovsky** [comment]
Yes, I was using iOS app to setup (but it was 3-4 weeks ago), and I think previously it was working well - I was able to send encrypted emails to this account.

**Comment by tomholub** [comment]
You were most likely in `SetupGenerateKeyViewController`, here (but the code has changed since a few weeks ago. But the functionality shouldn't have changed):

```swift
    func setupAccount(passPhrase: String,
                      storageMethod: StorageMethod,
                      viewController: ViewController) async throws {
        let userId = try getUserId()

        try await viewController.validateAndConfirmNewPassPhraseOrReject(passPhrase: passPhrase)

        let encryptedPrv = try await core.generateKey(
            passphrase: passPhrase,
            variant: .curve25519,
            userIds: [userId]
        )
        try await backupService.backupToInbox(keys: [encryptedPrv.key], for: user)

        keyStorage.addKeys(keyDetails: [encryptedPrv.key],
                           passPhrase: storageMethod == .persistent ? passPhrase: nil,
                           source: .generated,
                           for: user.email)

        if storageMethod == .memory {
            let passPhrase = PassPhrase(
                value: passPhrase,
                fingerprintsOfAssociatedKey: encryptedPrv.key.fingerprints
            )
            passPhraseService.savePassPhrase(with: passPhrase, storageMethod: .memory)
        }

        await submitKeyToAttesterAndShowAlertOnFailure(
            email: userId.email,
            publicKey: encryptedPrv.key.public,
            viewController: viewController
        )

        // sending welcome email is not crucial, so we don't handle errors
        _ = try? await attester.testWelcome(
            email: userId.email,
            pubkey: encryptedPrv.key.public
        )
    }
```

This is the part that should have happened:

```swift
        await submitKeyToAttesterAndShowAlertOnFailure(
            email: userId.email,
            publicKey: encryptedPrv.key.public,
            viewController: viewController
        )
```

and here 

```swift
    @discardableResult
    func updateKey(email: String, pubkey: String, token: String?) async throws -> String {
        let httpMethod: HTTPMetod
        let headers: [URLHeader]

        if let value = token {
            httpMethod = .post
            headers = [URLHeader(value: "Bearer \(value)", httpHeaderField: "Authorization")]
        } else {
            httpMethod = .put
            headers = []
        }

        let request = ApiCall.Request(
            apiName: Constants.apiName,
            url: urlPub(emailOrLongid: email),
            method: httpMethod,
            body: pubkey.data(),
            headers: headers
        )
        let res = try await ApiCall.asyncCall(request)
        return res.data.toStr()
    }
```

**Comment by tomholub** [comment]
To fix this manually, you can log into that account on browser extension, then go to settings -> additional -> Attester. There, it will allow you to re-submit the public key that's missing on Attester. (iOS app doesn't have this functionality yet)

**Comment by tomholub** [comment]
Or you could do it here from the mobile device. https://flowcrypt.com/attester/update-public-keys - paste your email and public key in (css is broken, it seems)

**Comment by sosnovsky** [comment]
But I signed in with my test account on another simulator and my public key was there.

If there was error during initial public key upload then it shouldn't appear on another device, right? Or it can be stored somewhere else?

**Comment by sosnovsky** [comment]
> I wonder if you had multiple accounts set up on a device that time? Contacts are shared across accounts after all. So if maybe one account saved its own public key into `Contacts`, then another account (but only on the save device) could have used it when sending?

Yes, I think there was second account configured on the device. Possibly it was on the same simulator where I had issues reading replies from you.

**Comment by tomholub** [comment]
> But I signed in with my test account on another simulator and my public key was there.
> If there was error during initial public key upload then it shouldn't appear on another device, right? Or it can be stored somewhere else?

When getting public keys of other users, the sources may be:
 - FlowCrypt Attester (most common)
 - WKD (if the recipient set it up, not your case)
 - Manual import (someone sends you a public key and you manually import it. iOS app doesn't support this yet)
 - Company-specific SKS server configured in OrgRules (not your case)

But when you're talking about your own public key, that one can be recreated/extracted from your own private key.

When you re-setup the account, it probably asked you for your pass phrase, which must have been the same as before. This is because it found your encrypted private key backup in inbox (we're talking about non-enterprise app behavior now). Once you gave it the pass phrase, it decrypted the private key, and it also extracted the public key from it.

I don't know, if at that point, the app saves your own extracted public key to contacts. If it did, there wouldn't be anything wrong about doing that.

**Comment by tomholub** [comment]
I don't see any code in the setup controllers that would be extracting your own public key from your private key and setting that in contacts. At least I don't see that on the iOS app. So I'm not sure how it got in your contacts before.

**Comment by sosnovsky** [comment]
Yes, app found 1 key backup during setup

![Simulator Screen Shot - iPhone 13 - 2021-11-09 at 16 54 26](https://user-images.githubusercontent.com/1062093/140947397-9c898554-8fa1-48cb-8e02-dd60c68ca828.png)

**Comment by tomholub** [comment]
One way to explain it would be that you did submit to Attester during setup, and then later manually did https://flowcrypt.com/attester/update-public-keys?intention=delete

**Comment by tomholub** [comment]
After you do that and go to contacts, does it show your public key there?

**Comment by tomholub** [comment]
The app crashes when I do that :-) #910

**Comment by tomholub** [comment]
I just tested on a blank app, where I went through setup (backup recovery) like the screen above. The public key of that account did not show in contacts.

So I'm really not sure how come it worked before for you.

One possible explanation could be #852 - if you managed to get public keys of another recipient recorded for this account due to that bug.

You could find out as follows:
 - look through emails on that account. Is there any email that was received from another account and is encrypted?
 - if yes, is it readable?

**Comment by sosnovsky** [comment]
> One way to explain it would be that you did submit to Attester during setup, and then later manually did https://flowcrypt.com/attester/update-public-keys?intention=delete

No, I didn't update public keys there.

App shows my own key on Settings -> Keys, not on Settings -> Contacts

**Comment by tomholub** [comment]
> App shows my own key on Settings -> Keys, not on Settings -> Contacts

Ok, that's expected behavior as currently implemented

**Comment by sosnovsky** [comment]
So public key is parsed from my private key and added to Settings -> Keys, right?

**Comment by tomholub** [comment]
> So public key is parsed from my private key and added to Settings -> Keys, right?

Yes. I mean a _keypair_ is saved to Settings -> Keys. A keypair is a pair of keys - both private and public, and they were both extracted from your backup. (OpenPGP private key automatically also includes Public Key information encoded in it, and so Public Key can be extracted from it, and saved along the private key in Settings -> Keys )

**Comment by tomholub** [comment]
But that doesn't explain why would the recipient turn green if you try to add it to recipients, even on the same device.

**Comment by sosnovsky** [comment]
> You could find out as follows:
> 
> * look through emails on that account. Is there any email that was received from another account and is encrypted?
> * if yes, is it readable?

I checked inbox and there is no received encrypted emails, only sent, so possibly this issue was from initial setup.
Then maybe we should have some check for this case when there was an issue sending pub key to attester? As it can be confusing to users that they're not able to receive encrypted emails even after setup.

**Comment by sosnovsky** [comment]
> But that doesn't explain why would the recipient turn green if you try to add it to recipients, even on the same device.

No, test account is colored gray when I try to send email to it from my `...@flowcrypt.com` account

**Comment by tomholub** [comment]
> > You could find out as follows:
> > 
> > * look through emails on that account. Is there any email that was received from another account and is encrypted?
> > * if yes, is it readable?
> 
> I checked inbox and there is no received encrypted emails, only sent, so possibly this issue was from initial setup. Then maybe we should have some check for this case when there was an issue sending pub key to attester? As it can be confusing to users that they're not able to receive encrypted emails even after setup.

Right now, you should see a modal when submission to Attester fails. But then if you tap the modal away, it won't be retried.

The modal could have two buttons: [retry] [skip]. So that you could at least retry it. That would be an improvement.

But because this is a consumer-app problem while my main focus these months is enterprise behavior, I'll file it for a future milestone to do maybe Q2 2022.

**Comment by tomholub** [comment]
In either case, I'm not submitting the app for consumer release on public app store until this is fixed. For now, just TestFlight beta (for consumers) and a real release for enterprise. That means, consumer-only issues are less important at the moment.

**Comment by tomholub** [comment]
Btw the bot will not bother you if you add emails that are `somewhere@example.com` or `somewhere@whatever.test`.

**Comment by sosnovsky** [comment]
Works well after I manually submitted key at https://flowcrypt.com/attester/update-public-keys, thanks for explanation!
